File: strategy.py
"""
strategy.py

Defines the rule-based strategy engine. A Strategy is a data structure
that decides, month by month, how much to invest given the current market
event and recent history. Every strategy in the project brief reduces to
a combination of these rules -- no strategy requires custom code, only
configuration. See presets.py for all 9 examples from the brief built
this way.
"""

from dataclasses import dataclass, field
from typing import Optional, List
from market_events import MarketEvent


@dataclass
class Trigger:
    """
    Describes the condition under which a rule fires.

    type:
        "event"             -> fires when the current month's MarketEvent
                                equals `event`
        "sequential_loss"   -> fires when there have been `streak` or more
                                consecutive losing months (Loss/ExtremeLoss/
                                Crash), including the current one
        "sequential_gain"   -> fires when there have been `streak` or more
                                consecutive gaining months (Gain/ExtremeGain/
                                Bubble), including the current one
        "return_threshold"  -> fires when the current month's return
                                satisfies `operator` against `value`
                                (operator in "lte", "lt", "gte", "gt")
        "drawdown_from_peak" -> fires when the current price is at or below
                                `value` (a negative fraction, e.g. -0.20)
                                relative to the highest price reached so
                                far in the simulation -- e.g. value=-0.20
                                fires once the index is 20%+ off its
                                running all-time high. Distinct from the
                                single-month event triggers: this tracks
                                the whole simulation's history, not just
                                the most recent month.
    """
    type: str
    event: Optional[MarketEvent] = None
    streak: Optional[int] = None
    operator: Optional[str] = None
    value: Optional[float] = None

    def matches(self, event: MarketEvent, monthly_return: float,
                loss_streak: int, gain_streak: int,
                drawdown_from_peak: float = 0.0) -> bool:
        if self.type == "event":
            return event == self.event
        if self.type == "sequential_loss":
            return loss_streak >= (self.streak or 1)
        if self.type == "sequential_gain":
            return gain_streak >= (self.streak or 1)
        if self.type == "return_threshold":
            ops = {
                "lte": lambda r, v: r <= v,
                "lt": lambda r, v: r < v,
                "gte": lambda r, v: r >= v,
                "gt": lambda r, v: r > v,
            }
            return ops[self.operator](monthly_return, self.value)
        if self.type == "drawdown_from_peak":
            return drawdown_from_peak <= self.value
        raise ValueError(f"Unknown trigger type: {self.type}")


@dataclass
class Action:
    """
    Describes what happens to the buy amount when a rule's trigger matches.

    type:
        "multiply"          -> buy = base_buy * value
        "set_fixed"         -> buy = value (use float('inf') for "buy as much as possible")
        "add_fixed"         -> buy = base_buy + value
        "skip"              -> buy = 0
        "scale_with_streak" -> buy = base_buy * multiplier, where the
                                multiplier grows with how long the current
                                streak has run:
                                    multiplier = start + increment * (streak - 1)
                                capped at `cap` if given. Pairs with a
                                "sequential_loss"/"sequential_gain" trigger
                                set to streak=1 (i.e. "active from the
                                first month of the streak onward"), so ONE
                                rule handles escalation instead of stacking
                                several fixed-multiplier rules by hand.
                                Example: start=1, increment=1, cap=5 gives
                                exactly "buy normally, then 2x, 3x, 4x...
                                capped at 5x" for a growing loss streak.
    """
    type: str
    value: Optional[float] = None
    start: float = 1.0     # only used by "scale_with_streak"
    increment: float = 0.0  # only used by "scale_with_streak"
    cap: Optional[float] = None  # only used by "scale_with_streak"

    def apply(self, base_buy: float, streak: int = 1) -> float:
        if self.type == "multiply":
            return base_buy * self.value
        if self.type == "set_fixed":
            return self.value
        if self.type == "add_fixed":
            return base_buy + self.value
        if self.type == "skip":
            return 0.0
        if self.type == "scale_with_streak":
            multiplier = self.start + self.increment * (streak - 1)
            if self.cap is not None:
                multiplier = min(multiplier, self.cap)
            return base_buy * multiplier
        raise ValueError(f"Unknown action type: {self.type}")


@dataclass
class Rule:
    trigger: Trigger
    action: Action


@dataclass
class DepletingBaseBuy:
    """
    Implements the "start high, then drop to income level" style strategy.
    While available cash allows buying `preferred_amount`, that amount is
    used as the base buy. Once available cash can no longer sustain it,
    the base buy falls back to `fallback_amount` (typically monthly income).
    """
    preferred_amount: float
    fallback_amount: float

    def compute(self, available_cash: float) -> float:
        return self.preferred_amount if available_cash >= self.preferred_amount \
            else self.fallback_amount


@dataclass
class Strategy:
    name: str
    rules: List[Rule] = field(default_factory=list)
    base_buy_override: Optional[DepletingBaseBuy] = None
    allow_selling: bool = False  # reserved for v2 (moving-average sell rules)
    description: str = ""  # short human-readable summary, shown alongside results

    def compute_buy_amount(self, base_buy: float, event: MarketEvent,
                            monthly_return: float, loss_streak: int,
                            gain_streak: int, available_cash: float,
                            drawdown_from_peak: float = 0.0) -> float:
        """
        Determine how much to invest this month.

        Every matching rule is applied IN ORDER, each one transforming the
        running buy amount before the next rule sees it -- rules compose
        rather than compete. This is what lets a per-event multiplier and
        a streak-escalation rule work together naturally: e.g. a "Loss"
        multiplier of 1.5x combined with a loss-streak escalation of
        +1x/month doesn't require choosing one or the other -- the
        escalation multiplies on top of the already-adjusted amount, so a
        3-month loss streak with both configured buys 1.5x * 3x = 4.5x the
        base. If no rule matches, the (possibly overridden) base buy is
        used unchanged.

        drawdown_from_peak: current price's fractional distance below the
            running all-time-high seen so far this simulation (always <= 0;
            0.0 means at a new peak). Defaults to 0.0 so callers/strategies
            that don't use "drawdown_from_peak" triggers are unaffected.
        """
        amount = base_buy
        if self.base_buy_override is not None:
            amount = self.base_buy_override.compute(available_cash)

        for rule in self.rules:
            if rule.trigger.matches(event, monthly_return, loss_streak, gain_streak,
                                     drawdown_from_peak=drawdown_from_peak):
                # Give the action whichever streak its trigger cares about,
                # so "scale_with_streak" can compute the right multiplier.
                # Defaults to 1 (no scaling effect) for triggers that
                # aren't streak-based at all.
                relevant_streak = 1
                if rule.trigger.type == "sequential_loss":
                    relevant_streak = loss_streak
                elif rule.trigger.type == "sequential_gain":
                    relevant_streak = gain_streak
                amount = rule.action.apply(amount, streak=relevant_streak)

        return max(0.0, amount)  # never invest a negative amount

File: js/strategy.js
/*
 * strategy.js
 *
 * The rule engine. A Strategy is data: a list of Rules, each pairing a
 * Trigger with an Action. All matching rules apply in order, each
 * transforming the running buy amount before the next rule sees it.
 */

const COMPARE_OPS = {
  lte: (a, b) => a <= b,
  lt: (a, b) => a < b,
  gte: (a, b) => a >= b,
  gt: (a, b) => a > b,
};

class Trigger {
  constructor({
    type,
    event = null,
    streak = null,
    operator = null,
    value = null,
    mode = null,
    children = null,
    period = null,
    fastPeriod = null,
    slowPeriod = null,
    direction = null,
    side = null,
  }) {
    this.type = type;
    this.event = event;
    this.streak = streak;
    this.operator = operator;
    this.value = value;
    // Only used by type "logic" -- a compound trigger combining several
    // child Triggers with AND ("all") / OR ("any") / NOT ("not", first
    // child only) semantics, built by the block-editor UI (see
    // block_editor.js). No preset uses one directly, so every other
    // trigger type is unaffected.
    this.mode = mode;
    this.children = children;
    // Only used by the price-indicator trigger types below ("ma_cross",
    // "ma_state", "ma_dual_cross", "ema_cross", "ema_state",
    // "rsi_threshold"). `period` is in MONTHS (this simulator has no
    // daily resolution).
    this.period = period;
    this.fastPeriod = fastPeriod;
    this.slowPeriod = slowPeriod;
    this.direction = direction;
    this.side = side;
  }

  matches(ctx) {
    switch (this.type) {
      case "event":
        return ctx.event === this.event;
      case "sequential_loss":
        return ctx.lossStreak >= (this.streak ?? 1);
      case "sequential_gain":
        return ctx.gainStreak >= (this.streak ?? 1);
      case "return_threshold":
        return COMPARE_OPS[this.operator](ctx.monthlyReturn, this.value);
      case "drawdown_from_peak":
        return ctx.drawdownFromPeak <= this.value;
      case "logic": {
        if (!this.children || this.children.length === 0) return false;
        if (this.mode === "not") return !this.children[0].matches(ctx);
        const results = this.children.map((c) => c.matches(ctx));
        return this.mode === "any" ? results.some(Boolean) : results.every(Boolean);
      }
      // --- price-indicator triggers (see indicators.js) ---
      case "price_threshold":
        return COMPARE_OPS[this.operator](ctx.price, this.value);
      case "ma_cross":
        return crossMatches(Indicators.sma(ctx.prices, this.period), ctx, this.direction);
      case "ma_state":
        return stateMatches(Indicators.sma(ctx.prices, this.period), ctx, this.side);
      case "ema_cross":
        return crossMatches(Indicators.ema(ctx.prices, this.period), ctx, this.direction);
      case "ema_state":
        return stateMatches(Indicators.ema(ctx.prices, this.period), ctx, this.side);
      case "ma_dual_cross":
        return dualCrossMatches(
          Indicators.sma(ctx.prices, this.fastPeriod),
          Indicators.sma(ctx.prices, this.slowPeriod),
          ctx,
          this.direction
        );
      case "rsi_threshold": {
        const v = Indicators.rsi(ctx.prices, this.period)[ctx.monthIndex];
        if (v === null) return false;
        return COMPARE_OPS[this.operator](v, this.value);
      }
      default:
        throw new Error(`Unknown trigger type: ${this.type}`);
    }
  }
}

// price vs. a single indicator series crossing this month
function crossMatches(series, ctx, direction) {
  const i = ctx.monthIndex;
  if (i < 1) return false;
  const cur = series[i], prev = series[i - 1];
  if (cur === null || prev === null) return false;
  const prevPrice = ctx.prices[i - 1], curPrice = ctx.prices[i];
  return direction === "up" ? prevPrice <= prev && curPrice > cur : prevPrice >= prev && curPrice < cur;
}

// price currently above/below a single indicator series
function stateMatches(series, ctx, side) {
  const cur = series[ctx.monthIndex];
  if (cur === null) return false;
  return side === "above" ? ctx.price > cur : ctx.price < cur;
}

// two indicator series (fast vs. slow) crossing this month
function dualCrossMatches(fastSeries, slowSeries, ctx, direction) {
  const i = ctx.monthIndex;
  if (i < 1) return false;
  const fCur = fastSeries[i], fPrev = fastSeries[i - 1];
  const sCur = slowSeries[i], sPrev = slowSeries[i - 1];
  if (fCur === null || fPrev === null || sCur === null || sPrev === null) return false;
  return direction === "up" ? fPrev <= sPrev && fCur > sCur : fPrev >= sPrev && fCur < sCur;
}

class Action {
  constructor({ type, value = null, start = 1.0, increment = 0.0, cap = null }) {
    this.type = type;
    this.value = value;
    this.start = start;
    this.increment = increment;
    this.cap = cap;
  }

  apply(baseBuy, streak = 1) {
    switch (this.type) {
      case "multiply":
        return baseBuy * this.value;
      case "set_fixed":
        return this.value;
      case "add_fixed":
        return baseBuy + this.value;
      case "skip":
        return 0.0;
      case "scale_with_streak": {
        let multiplier = this.start + this.increment * (streak - 1);
        if (this.cap !== null && this.cap !== undefined) {
          multiplier = Math.min(multiplier, this.cap);
        }
        return baseBuy * multiplier;
      }
      default:
        throw new Error(`Unknown action type: ${this.type}`);
    }
  }
}

class Rule {
  constructor(trigger, action) {
    this.trigger = trigger;
    this.action = action;
  }
}

// "Start high, then drop to income level" style base-buy override.
class DepletingBaseBuy {
  constructor(preferredAmount, fallbackAmount) {
    this.preferredAmount = preferredAmount;
    this.fallbackAmount = fallbackAmount;
  }

  compute(availableCash) {
    return availableCash >= this.preferredAmount ? this.preferredAmount : this.fallbackAmount;
  }

  clone() {
    return new DepletingBaseBuy(this.preferredAmount, this.fallbackAmount);
  }
}

class Strategy {
  constructor({ name, rules = [], baseBuyOverride = null, allowSelling = false, description = "" }) {
    this.name = name;
    this.rules = rules;
    this.baseBuyOverride = baseBuyOverride;
    this.allowSelling = allowSelling;
    this.description = description;
  }

  // `ctx` carries everything a Trigger might need to evaluate itself:
  // {event, monthlyReturn, lossStreak, gainStreak, drawdownFromPeak,
  //  availableCash, prices, monthIndex, price} -- see simulator.js, which
  // builds it fresh each month.
  computeBuyAmount(baseBuy, ctx) {
    let amount = baseBuy;
    if (this.baseBuyOverride !== null) {
      amount = this.baseBuyOverride.compute(ctx.availableCash);
    }

    for (const rule of this.rules) {
      if (rule.trigger.matches(ctx)) {
        let relevantStreak = 1;
        if (rule.trigger.type === "sequential_loss") relevantStreak = ctx.lossStreak;
        else if (rule.trigger.type === "sequential_gain") relevantStreak = ctx.gainStreak;
        amount = rule.action.apply(amount, relevantStreak);
      }
    }

    return Math.max(0.0, amount);
  }

  // Each backtest run needs its own deep copy, so a stateful
  // DepletingBaseBuy's patched fallback never leaks across runs.
  clone() {
    return new Strategy({
      name: this.name,
      description: this.description,
      allowSelling: this.allowSelling,
      rules: this.rules.map(
        (r) => new Rule(new Trigger({ ...r.trigger }), new Action({ ...r.action }))
      ),
      baseBuyOverride: this.baseBuyOverride ? this.baseBuyOverride.clone() : null,
    });
  }
}
